src/binding/c/request_api.txt

```diff
@@ -0,0 +1,617 @@
+            request_ptrs[i] = NULL;
+            n_inactive += 1;
+        }
+    }
+
+    if (n_inactive == incount) {
+        *outcount = MPI_UNDEFINED;
+        goto fn_exit;
+    }
+
+    mpi_errno = MPID_Testsome(incount, request_ptrs, outcount, array_of_indices, array_of_statuses);
+    /* --BEGIN ERROR HANDLING-- */
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+    /* --END ERROR HANDLING-- */
+
+    if (proc_failure)
+        mpi_errno = MPI_ERR_IN_STATUS;
+
+    if (*outcount == MPI_UNDEFINED)
+        goto fn_exit;
+
+    for (int i = 0; i < *outcount; i++) {
+        int idx = array_of_indices[i];
+        MPI_Status *status_ptr = (array_of_statuses == MPI_STATUSES_IGNORE) ?
+            MPI_STATUS_IGNORE : &array_of_statuses[i];
+        int rc = MPIR_Request_completion_processing(request_ptrs[idx], status_ptr);
+        if (!MPIR_Request_is_persistent(request_ptrs[idx])) {
+            MPIR_Request_free(request_ptrs[idx]);
+            array_of_requests[idx] = MPI_REQUEST_NULL;
+        }
+        if (rc == MPI_SUCCESS) {
+            request_ptrs[idx] = NULL;
+        } else {
+            mpi_errno = MPI_ERR_IN_STATUS;
+            if (status_ptr != MPI_STATUS_IGNORE) {
+                status_ptr->MPI_ERROR = rc;
+            }
+        }
+    }
+
+    if (mpi_errno == MPI_ERR_IN_STATUS) {
+        if (array_of_statuses != MPI_STATUSES_IGNORE) {
+            for (int i = 0; i < *outcount; i++) {
+                if (request_ptrs[array_of_indices[i]] == NULL) {
+                    array_of_statuses[i].MPI_ERROR = MPI_SUCCESS;
+                }
+            }
+        }
+    }
+}
+
+MPI_Wait:
+    .desc: Waits for an MPI request to complete
+    .docnotes: waitstatus
+{
+    mpi_errno = MPIR_Wait(request, status);
+    if (mpi_errno)
+        goto fn_fail;
+}
+
+MPI_Waitall:
+    .desc: Waits for all given MPI Requests to complete
+    .docnotes: waitstatus
+/*
+    Notes:
+
+    If one or more of the requests completes with an error, 'MPI_ERR_IN_STATUS' is
+    returned.  An error value will be present is elements of 'array_of_statuses'
+    associated with the requests.  Likewise, the 'MPI_ERROR' field in the status
+    elements associated with requests that have successfully completed will be
+    'MPI_SUCCESS'.  Finally, those requests that have not completed will have a
+    value of 'MPI_ERR_PENDING'.
+
+    While it is possible to list a request handle more than once in the
+    array_of_requests, such an action is considered erroneous and may cause the
+    program to unexecpectedly terminate or produce incorrect results.
+*/
+{ -- error_check -- array_of_statuses
+    if (count > 0) {
+        MPIR_ERRTEST_ARGNULL(array_of_statuses, "array_of_statuses", mpi_errno);
+    }
+}
+{
+    mpi_errno = MPIR_Waitall(count, array_of_requests, array_of_statuses);
+    if (mpi_errno)
+        goto fn_fail;
+}
+
+MPI_Waitany:
+    .desc: Waits for any specified MPI Request to complete
+    .docnotes: waitstatus
+/*
```

Comment:
in waitany, `comm_ptr` is declared but is not set to request_ptr->comm

Reply:
It was an error to declare `comm_ptr` in the case of request array such as in `MPI_Waitany` as it is not clear which communicator from which request should be used.

The leaked `comm_ptr` declare is now fixed.